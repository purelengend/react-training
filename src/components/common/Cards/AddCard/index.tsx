import { ReactNode } from 'react';
import plusIcon from '../../../../assets/icons/plus-icon.svg';
import addCardStyles from './add-card.module.css';
interface AddCardProps {
  children: ReactNode;
}
export const AddCard = ({children}: AddCardProps) => {
  return (
    <div
      id="add-card"
      className={`d-flex-center ${addCardStyles['addition-card']}`}
    >
      <div className={`d-flex-col ${addCardStyles['addition-wrapper']}`}>
        <img src={plusIcon} alt="Plus Icon" className="primary-icon" />
        <p className={addCardStyles['addition-content']}>{children}</p>
      </div>
    </div>
  );
};
