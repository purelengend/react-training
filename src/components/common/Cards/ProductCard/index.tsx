import productCardStyles from '@components/common/Cards/ProductCard/product-card.module.css';
import deleteIcon from '@assets/icons/cross-icon.svg';
import editIcon from '@assets/icons/edit-icon.svg';
import { Button } from '@components/common/Button';
import isEqual from 'react-fast-compare';
import { memo } from 'react';

export interface Food {
  id: string;
  name: string;
  price: number;
  quantity: number;
  imageUrl: string;
  createdAt: Date;
}
interface ProductCardProps {
  product: Food;
  onDeleteClick: React.MouseEventHandler<HTMLButtonElement>;
  onEditClick: React.MouseEventHandler<HTMLButtonElement>;
}
export const ProductCard = memo(
  ({ product, onDeleteClick, onEditClick }: ProductCardProps) => {
    return (
      <div
        className={`d-flex-center d-flex-col ${productCardStyles['product-card']}`}
      >
        <Button
          onClick={onDeleteClick}
          className={`d-flex-center ${productCardStyles['secondary-icon']} ${productCardStyles['delete-btn']}`}
        >
          <img
            width="20"
            height="20"
            src={deleteIcon}
            alt="Cross Icon"
            data-id={product.id}
          />
        </Button>

        <div className={`d-flex-col ${productCardStyles['product-wrapper']}`}>
          <img
            width="127"
            height="127"
            src={product.imageUrl}
            alt={product.name}
            className={productCardStyles['primary-product']}
          />
          <div
            className={`d-flex-col ${productCardStyles['product-content-wrapper']}`}
          >
            <p className={productCardStyles['product-name']}>{product.name}</p>
            <div
              className={`d-flex-center ${productCardStyles['product-detail']}`}
            >
              $ {product.price}
              <div className={productCardStyles.separate}></div>
              {product.quantity} Bowls
            </div>
          </div>
        </div>

        <Button
          className={`d-flex-center ${productCardStyles['product-mutation']} ${productCardStyles.mutation}`}
          onClick={onEditClick}
        >
          <img
            src={editIcon}
            alt="Edit Icon"
            className={`primary-icon ${productCardStyles.mutation}`}
          />
          <p
            className={`${productCardStyles['mutation-content']} ${productCardStyles.mutation}`}
          >
            Edit dish
          </p>
        </Button>
      </div>
    );
  },
  isEqual
);

ProductCard.whyDidYouRender = true;
