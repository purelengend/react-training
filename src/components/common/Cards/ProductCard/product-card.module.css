.product-card {
  width: 246px;
  height: 299px;
  border-radius: var(--border-primary);
  border: 1px solid var(--base-dark-line);
  position: relative;
  gap: 16px;
  transition: box-shadow 0.2s ease-in-out;
}

.product-wrapper {
  align-items: center;
  gap: 14px;
  text-align: center;
}

.primary-product {
  width: 127px;
  height: 127px;
  object-fit: cover;
  border-radius: 50%;
}

.product-content-wrapper {
  min-width: 144px;
  max-width: fit-content;
  gap: 8px;
  font-size: var(--font-size-xs);
  color: var(--text-light);
  font-weight: var(--font-weight-normal);
  line-height: var(--line-height-lg);
}

.secondary-icon {
  width: 20px;
  height: 20px;
  position: absolute;
  top: 16px;
  right: 16px;
  opacity: 0;
  transition: all 0.3s ease-in-out;
}

.product-card:hover .secondary-icon {
  opacity: 1;
}

.product-name {
  min-height: 41px;
  overflow: hidden;
  text-overflow: ellipsis;
  color: var(--text-primary);
  font-size: var(--font-size-xs);
  font-weight: var(--font-weight-medium);
  line-height: var(--line-height-md);
}

.product-detail {
  gap: 8px;
}

.separate {
  width: 4px;
  height: 4px;
  border-radius: 50%;
  background-color: var(--text-light);
}

.product-mutation {
  width: 246px;
  height: 52px;
  background-color: rgba(234, 124, 105, 0.24);
  border: 1px solid var(--alternative-color);
  border-top: none;
  border-bottom-left-radius: var(--border-primary);
  border-bottom-right-radius: var(--border-primary);
  margin-bottom: -18px;
  gap: 8px;
}

.mutation-content {
  color: var(--primary-color);
  font-size: var(--font-size-xs);
  font-weight: var(--font-weight-bold);
  line-height: var(--line-height-lg);
}

.delete-btn {
  background-color: transparent;
  border: 0;
  padding: 0;
}
.delete-btn:hover {
  cursor: pointer;
}

.product-card:hover {
  box-shadow: 0px 8px 24px 0px var(--box-shadow-primary);
}
