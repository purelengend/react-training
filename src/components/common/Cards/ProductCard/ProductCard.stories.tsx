import type { Meta, StoryObj } from '@storybook/react';
import { ProductCard } from '@components/common/Cards/ProductCard';
import '@components/common/Cards/ProductCard/product-card.module.css';
const meta: Meta<typeof ProductCard> = {
  title: 'Example/ProductCard',
  component: ProductCard,
  decorators: [
    Story => (
      <div
        style={{
          display: 'grid',
          placeItems: 'center',
          height: '90vh'
        }}
      >
        <Story />
      </div>
    )
  ],
  argTypes: {
    onEditClick: {
      action: 'clicked',
      description: 'this is the mock click function handler'
    },
    onDeleteClick: {
      action: 'clicked',
      description: 'this is the mock click function handler'
    },
    product: {
      description: 'the product object contains all the information fields'
    }
  },
  parameters: {
    controls: {
      expanded: true
    }
  }
};

export default meta;

type Story = StoryObj<typeof ProductCard>;

export const ProductCardExample: Story = {
  args: {
    product: {
      id: '0',
      name: 'test product',
      price: 123,
      imageUrl:
        'https://images.unsplash.com/photo-1614777986387-015c2a89b696?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=3774&q=80',
      quantity: 4,
      createdAt: new Date()
    }
  }
};
