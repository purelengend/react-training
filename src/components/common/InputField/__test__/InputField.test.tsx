import { Food } from '@components/common/Cards/ProductCard';
import { customRenderer, renderHook } from '@src/test/test-utils';
import { useForm } from 'react-hook-form';

import { InputField, InputFieldProps } from '..';

describe('InputField test cases', () => {
  const {
    result: {
      current: { register }
    }
  } = renderHook(() => useForm<Food>());
  const mockInputProps: InputFieldProps = {
    label: 'mock',
    labelClass: 'mock-label',
    htmlFor: 'mock',
    inputClass: 'mock-input',
    placeholder: 'mock placeholder',
    type: 'text',
    name: 'name',
    register: register,
    typeValue: 'valueAsNumber'
  };

  it('should render correctly', () => {
    const mockInputField = customRenderer(<InputField {...mockInputProps} />);

    expect(mockInputField.toJSON()).toMatchSnapshot();
  });
});
