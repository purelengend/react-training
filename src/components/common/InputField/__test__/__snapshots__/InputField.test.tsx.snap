// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputField test cases > should render correctly 1`] = `
[
  <label
    className="mock-label"
    htmlFor="mock"
  >
    mock
  </label>,
  <input
    className="mock-input"
    id="mock"
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="mock placeholder"
    step="any"
    type="text"
  />,
]
`;
